crawler.py

```diff
@@ -265,3 +266,62 @@ def run_all(driver, data_folder):
 
     for filename in ['code-coverage-gcda.zip', jsvm_output_file]:
         os.remove(filename)
+
+    with open('{}/diff.json'.format(data_folder), 'r') as report:
+        parsed_json = json.load(report)
+
+    file_obj = open('{}/report.info'.format(data_folder), 'w')
+
+    source_files = parsed_json['source_files']
+
+    file_obj.write('TN\n')
+    for source_file in source_files:
+        file_obj.write('SF:{}\n'.format(source_file['name']))
+        executed = 0
+
+        # Functions
+        if len(source_file['functions']) != 0:
+            for function in source_file['functions']:
+                file_obj.write('FN:{},{}\n'.format(function['start'], function['name']))
+            for function in source_file['functions']:
+                if function['exec'] is True:
+                    file_obj.write('FNDA:{},{}\n'.format(1, function['name']))
+                    executed += 1
+                else:
+                    file_obj.write('FNDA:{},{}\n'.format(0, function['name']))
+        file_obj.write('FNF:{}\n'.format(len(source_file['functions'])))
+        file_obj.write('FNH:{}\n'.format(executed))
+
+        # Branches
+        branch_hits = 0
+        if len(source_file['branches']) != 0:
+            for branch in source_file['branches']:
+                if branch['taken'] is True:
+                    file_obj.write('BRDA:{},0,{},{}\n'.format(branch['line'], branch['number'], 1))
+                    branch_hits += 1
+                else:
+                    file_obj.write('BRDA:{},0,{},{}\n'.format(branch['line'], branch['number'], '-'))
+        file_obj.write('BRF:{}\n'.format(len(source_file['branches'])))
+        file_obj.write('BRH:{}\n'.format(branch_hits))
+
+        # Lines
+        line_number = 0
+        cov_line_count = 0
+        if len(source_file['coverage']) != 0:
+            for line in source_file['coverage']:
+                line_number += 1
+                if line is not None:
+                    file_obj.write('DA:{},{}\n'.format(line_number, line))
+                    if line > 0:
+                        cov_line_count += 1
+        file_obj.write('LF:{}\n'.format(len(source_file['coverage'])))
+        file_obj.write('LH:{}\n'.format(cov_line_count))
+        file_obj.write('end_of_record\n')
+    file_obj.close()
+
+    codecoverage.download_genhtml()
+
+    ret = subprocess.call(['lcov-bin/usr/local/bin/genhtml', '-o', os.path.join(data_folder, 'report'), '--show-details', '--highlight', '--ignore-errors', 'source', '--legend', os.path.join(data_folder, 'report.info'), '--prefix', 'tools/firefox'])
```

Comment:
You could use the `generate_html_report` function of `codecoverage`.

Reply:
And you'll need to clone the Firefox source code repository (the "tools/firefox" directory doesn't contain the source files, right?)

Reply:
`generate_html_report` will put the generated report in the current working directory rather than in the directory where the `diff.json` and data files are.

Reply:
It means I need to clone it from Mercurial, yep? It contains coverage build of Firefox. But I used it previously as input with source files for firefox-code-coverage and it worked. :) Now, the report was generated too.

Reply:
> generate_html_report will put the generated report in the current working directory rather than in the directory where the diff.json and data files are.

We can add a parameter to generate_html_report to choose the output directory.

> It means I need to clone it from Mercurial, yep? It contains coverage build of Firefox. But I used it previously as input with source files for firefox-code-coverage and it worked. :) Now, the report was generated too.

Does the report contain the source code? I would expect it to work, but with the HTML not having the source code information.

Reply:
It doesn't: only line numbers with execution counts without code. But with the cloned Firefox source code passed as `src_dir` I get the same result.  

I will create PR with modifications to `generate_html_report` shortly.

Reply:
> It doesn't: only line numbers with execution counts without code. But with the cloned Firefox source code passed as src_dir I get the same result.

We should actually get the source files shown in the HTML report. Can you try running genhtml manually? What's the output like?

Reply:
I've tried running `codecoverage.py` again with clonned source files and it worked for Firefox coverage report: the source code is shown in the report. 
It looks much clearer now. I was comparing manually the line numbers in coverage report with source code in Mercurial previously. :p Thanks for the correction.

Reply:
Awesome!